packages/babel-parser/src/plugins/flow.js

```diff
@@ -2839,4 +2877,418 @@ export default (superClass: Class<Parser>): Class<Parser> =>
+          `before in enum \`${enumName}\`.`,
+      );
+    }
+
+    flowEnumErrorInconsistentMemberValues(
+      pos: number,
+      { enumName }: { enumName: string },
+    ): void {
+      this.raise(
+        pos,
+        `Enum \`${enumName}\` has inconsistent member initializers. Either use no initializers, or ` +
+          `consistently use literals (either booleans, numbers, or strings) for all member initializers.`,
+      );
+    }
+
+    flowEnumErrorInvalidExplicitType(
+      pos: number,
+      {
+        enumName,
+        suppliedType,
+      }: { enumName: string, suppliedType: null | string },
+    ): void {
+      const suggestion =
+        `Use one of \`boolean\`, \`number\`, \`string\`, or \`symbol\` in ` +
+        `enum \`${enumName}\`.`;
+      const message =
+        suppliedType === null
+          ? `Supplied enum type is not valid. ${suggestion}`
+          : `Enum type \`${suppliedType}\` is not valid. ${suggestion}`;
+      this.raise(pos, message);
+    }
+
+    flowEnumErrorInvalidMemberInitializer(
+      pos: number,
+      { enumName, explicitType, memberName }: EnumContext,
+    ): void {
+      let message = null;
+      switch (explicitType) {
+        case "boolean":
+        case "number":
+        case "string":
+          message =
+            `Enum \`${enumName}\` has type \`${explicitType}\`, so the initializer of ` +
+            `\`${memberName}\` needs to be a ${explicitType} literal.`;
+          break;
+        case "symbol":
+          message =
+            `Symbol enum members cannot be initialized. Use \`${memberName},\` in ` +
+            `enum \`${enumName}\`.`;
+          break;
+        default:
+          // null
+          message =
+            `The enum member initializer for \`${memberName}\` needs to be a literal (either ` +
+            `a boolean, number, or string) in enum \`${enumName}\`.`;
+      }
+      this.raise(pos, message);
+    }
+
+    flowEnumErrorNumberMemberNotInitialized(
+      pos: number,
+      { enumName, memberName }: { enumName: string, memberName: string },
+    ): void {
+      this.raise(
+        pos,
+        `Number enum members need to be initialized, e.g. \`${memberName} = 1\` in enum \`${enumName}\`.`,
+      );
+    }
+
+    flowEnumErrorStringMemberInconsistentlyInitailized(
+      pos: number,
+      { enumName }: { enumName: string },
+    ): void {
+      this.raise(
+        pos,
+        `String enum members need to consistently either all use initializers, or use no initializers, ` +
+          `in enum \`${enumName}\`.`,
+      );
+    }
+
+    flowEnumMemberInit(): EnumMemberInit {
+      const startPos = this.state.start;
+      const endOfInit = () => this.match(tt.comma) || this.match(tt.braceR);
+      switch (this.state.type) {
+        case tt.num: {
+          const literal = this.parseLiteral(this.state.value, "NumericLiteral");
```

Comment:
Not related to this PR but I am curious why
```js
enum E of number { A = 0 }
```
is valid but
```js
enum E of number { A = -0 }
```
is not valid.

Reply:
Not a super strong reason, other than: 
- As a general principle, we're keeping the initial version as restrictive as possible while still being useful, with the idea that you can always allow more behaviour in the future if there is some compelling use-case, but it is difficult to make things more restrictive
- When talking to potential users, no one mentioned wanting negative numbers as initializers
- It simplifies the implementation for v0 of Flow enums

I can imagine a future iteration of Flow enums where we allow arbitrary expressions as initializers, as long as they can be statically evaluated to a number/string/boolean, which would allow for this (and other more complicated, but still statically evaluable, expressions)